arch/x86/core/pcie.c

```diff
@@ -206,22 +206,19 @@ uint32_t pcie_msi_map(unsigned int irq,
 uint16_t pcie_msi_mdr(unsigned int irq,
 		      msi_vector_t *vector)
 {
-#ifdef CONFIG_PCIE_MSI_X
-	if ((vector != NULL) && (vector->msix)) {
-		return 0x4000U | vector->arch.vector;
-	}
-#endif
+	uint16_t msg_data_reg = 0;
 
-	if (vector == NULL) {
+#if !defined(CONFIG_INTEL_VTD_ICTL)
+	if (vector != NULL) {
+		msg_data_reg = vector->arch.vector;
+	} else {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
```

Comment:
this is the default code branch executed if neither MSI-X nor Multi-MSI capabilities are available and it is assumed that this branch is used only if VT-D is enabled. But it is also executed when only single vector MSI should be used. This would lead in a failure as the MSI vector used to trigger an IRQ event will be wrong.